### PR TITLE
Make _findMany only fetch unique IDs

### PR DESCRIPTION
**Pull request by orf**
This is a MR to address #4367. I think it is as simple as this, I will amend this and write some tests if the issue isn't rejected.

### CODE REVIEW

**Comment by bmac** [comment]
Looks like a good change @orf. Once you add a test I think we can merge this pr.

**Comment by orf** [comment]
Done @bmac, is that test sufficient?

**Comment by orf** [comment]
Done :+1:

**Comment by pangratz** [comment]
It looks like the test doesn't fail when the changes in `finders.js` are reverted. I am curious how it is possible that multiple find requests are triggered in the first place. Can you re-create a use case which demonstrates the issue you have when you reported #4367?

Don't get me wrong, in general I agree that this change is ok, I just want to find out what the underlying issue is here...

**Comment by orf** [comment]
I'm not entirely sure what causes it, but it definitely happens occasionally. I just set a breakpoint in my app to break when this occurs, and it just did with this:

`[04/May/2016 16:16:46] "GET /api/employees/?ids%5B%5D=EMPLOYEE_NAME&ids%5B%5D=EMPLOYEE_NAME HTTP/1.1" 200 194`

**Comment by fivetanley** [comment]
> It looks like the test doesn't fail when the changes in finders.js are reverted

Maybe this is because we are using `.finally` in the tests? I think we just want to return the promise instead. That will fail if the promise rejects, and pass if it succeeds.

**Comment by nadavshatz** [comment]
I've personally encountered it when a model was being loaded by a route and then an ember-concurrency task that is set to run in an inner component was trying to reload said model.
it would create the same issue.

and because the reload would return a json with root "plural" it would not work and the reload would fail.

**Comment by pangratz** [comment]
@fivetanley I think using `finally` is fine. What I mean is that the `assert.deepEqual(ids, ["1", "2"], 'Correct unique ids were passed in to findMany');` doesn't fail when the change is reverted, so it looks like the current test doesn't cover the bug which is solved here.

@orf are you using custom `primaryKey`s in your serializers? If you can reliably trigger this bug in a specific view you have, are you interested in pairing so we can dig into this?

@nadavshatz can you share some code?

**Comment by nadavshatz** [comment]
@pangratz http://emberjs.jsbin.com/pedijepepu/edit?html,js,output
Finally got a version working!

Take a look at the console at the mockjax log and see that the `findRecord` from the route is combined with the `reload` from the component into one call with the same id showing twice.

**Comment by pangratz** [comment]
Thanks @nadavshatz, this helped a ton tracking down the issue. @orf also thanks for your test case, you we're very close: the bug can be reproduced if a record is already pushed into the store and it is fetched multiple times within the same run-loop (either via a reload or a background reload).

@orf can you confirm that the record(s) is/are already pushed to the store when you run into the issue reported in #4367? I currently don't see a way how this bug occurs for multiple `findRecord`'s within the same run-loop, without the record(s) being already in the store beforehand.

---

The root of the bug basically lies within [`store.scheduleFetch`](https://git.io/vrGnq), where it is checked if there is already a `_loadingPromise` for the `InternalModel` [here](https://git.io/vrGnm), and if there is none, then another fetch item is created, without checking if there might already be a fetch scheduled for the given `InternalModel`.

To fix this bug there are several different approaches:

- looks like the state machine currently doesn't always set the `_loadingPromise` for the `InternalModel` when the [`loadingData`](https://git.io/vrGnz) event is triggered (compare https://git.io/vrGni to https://git.io/vrGnw). As pointed out in the comment above that event handler, I am not sure what implications there are when we adapt the state machine here.
- another fix could be to check if `this._pendingFetch` already contains [an item for the `internalModel`](https://git.io/vrGny) and if so, return its' resolvers promise.

One of those approaches would be suitable as a quick fix here IMHO.

---

On the long run and given that there are already some bugs around the find coalescing stuff (#4190, #4367) and there is need for bulk saving #1849, I would suggest to resurrect @mike-north's #2891 and refactor the whole coalescing stuff a whole. But before that happens this should definitely be fixed in the meantime.

@emberjs/ember-data-contributors thoughts?

File: addon/-private/system/store/finders.js [hunk 1]
```diff
@@ -55,7 +55,8 @@ export function _find(adapter, store, typeClass, id, internalModel, options) {
 
 export function _findMany(adapter, store, typeClass, ids, internalModels) {
   let snapshots = Ember.A(internalModels).invoke('createSnapshot');
-  let promise = adapter.findMany(store, typeClass, ids, snapshots);
+  let unique_ids = Ember.A(ids).uniq();
```
Comment (sandstrom) [inline addon/-private/system/store/finders.js]: Should this be `uniqueIds`? Variable names are typically camelCased in Javascript.

**Comment by broerse** [comment]
It seems unloadRecord should run for the first id. Just as in this test:  https://github.com/jlami/ember-bug-empty-data

**Comment by broerse** [comment]
See also https://github.com/emberjs/data/issues/4463

**Comment by stefanpenner** [comment]
Just reviewing the code which invokes `_findMany` im not sure it works correctly with duplicate `ids` as `seeking` within `_flushPendingFetchForType` will only store the last deferred and resolve that, not an array of all reserves.

I believe more work is required to make this feature complete, or someone should demonstrate ^ is already handled.

**Comment by runspired** [comment]
Closing in favor of #4596 test-pr and eventual fix for fetch scheduling

**Pull request closed without merging**